src/main/kotlin/com/intellij/ml/llm/template/utils/CodeTransformer.kt

```diff
@@ -0,0 +1,115 @@
+package com.intellij.ml.llm.template.utils
+
+import com.intellij.lang.java.JavaLanguage
+import com.intellij.ml.llm.template.customextractors.MyInplaceExtractionHelper
+import com.intellij.ml.llm.template.extractfunction.EFCandidate
+import com.intellij.ml.llm.template.models.FunctionNameProvider
+import com.intellij.ml.llm.template.models.MyMethodExtractor
+import com.intellij.openapi.diagnostic.Logger
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.editor.ex.EditorEx
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import com.intellij.psi.util.PsiTreeUtil
+import org.jetbrains.kotlin.idea.KotlinLanguage
+import org.jetbrains.kotlin.idea.refactoring.introduce.extractFunction.ExtractKotlinFunctionHandler
+import org.jetbrains.kotlin.utils.addIfNotNull
+
+class CodeTransformer : Observable {
+    private val logger = Logger.getInstance("#com.intellij.ml.llm")
+    private val observers = mutableListOf<Observer>()
+
+    fun applyCandidate(efCandidate: EFCandidate, project: Project, editor: Editor, file: PsiFile): Boolean {
+        var applicationResult = EFApplicationResult.OK
+        var reason = ""
+
+        if (!isCandidateValid(efCandidate)) {
+            applicationResult = EFApplicationResult.FAIL
+            reason = "invalid extract function candidate"
+        } else {
+            editor.selectionModel.setSelection(efCandidate.offsetStart, efCandidate.offsetEnd)
+            try {
+                invokeExtractFunction(efCandidate.functionName, project, editor, file)
+            } catch (e: Exception) {
+                applicationResult = EFApplicationResult.FAIL
+                reason = e.message ?: ""
+            } catch (e: Error) {
+                applicationResult = EFApplicationResult.FAIL
+                reason = e.message ?: ""
+            }
+        }
+
+        notifyObservers(
+            EFNotification(
+                result = applicationResult,
+                reason = reason,
+                candidate = efCandidate
+            )
+        )
+        return applicationResult == EFApplicationResult.OK
+    }
+
+    private fun invokeExtractFunction(newFunctionName: String, project: Project, editor: Editor?, file: PsiFile?) {
+        val functionNameProvider = FunctionNameProvider(newFunctionName)
+        when (file?.language) {
+            JavaLanguage.INSTANCE -> {
+                MyMethodExtractor.invokeOnElements(
+                    project, editor, file, findSelectedPsiElements(editor, file), FunctionNameProvider(newFunctionName)
+                )
+            }
+
+            KotlinLanguage.INSTANCE -> {
+                val dataContext = (editor as EditorEx).dataContext
+                val allContainersEnabled = false
```

Comment:
Could you please clarify what `allContainersEnabled` variable means? What containers?

Reply:
I'm not sure what `allContainersEnabled` means exactly. It comes from package `org.jetbrains.kotlin.idea.refactoring.introduce.extractFunction.ExtractKotlinFunctionHandler` it's the first parameter and it's being initialized to false by default. There is no documentation in the source file. I kept it false too.